### ISSUE TITLE
Non line based TokensProvider

### ISSUE THREAD

**Issue by dhakehurst** [microsoft/monaco-editor#564]
I want to provide a TokenProvide to the monaco editor.
However my list of tokens defines a start and end (or start and length) offset from the start of the text in the editor.

I.e. I am not tokenising each line separately.

I cannot see how to provide tokens in this manner. The IModel of the editor seems to expect a line by line set of tokens.

Is it possible to provide a Model that is not line based? 
(I believe some other js editors do provide this option)

**Comment by miho** [comment]
I think this feature is actually very important for semantic highlighting. I understand that a simple colorizer is much more performance friendly. But there should be an option for it. I hope it will be added in a future release. Thanks.

**Comment by piersdeseilligny** [comment]
@alexdima is there any timeline for when this might be implemented? I am attempting to implement a tokenizer with quite specific behaviour, where for example, in the following situation:

```
1:   Line 1
2:   Line 2
3:
4:   Line 4 
5:   Line 5
6:   Line 6
7:
8:
9:   Line 9
```

A change on line 1, will cause a retokenization on lines 1-3, a change on line 5 will cause a retokenization on lines 3-8, a change on line 9 will cause a retokenziation on lines 7-9, etc.... This is because the tokenization of a given line might vary based on what happens until the next empty line.

So basically, I'm looking to tokenize the text by chunks delimited with empty lines. I have a vague prototype working with decorations, but for my usecase there's just no way of using decorations or semantic tokens, mainly for performance reasons.

Is there any realistic way of doing this currently, even if it's hacky and/or involves fiddling with some unexposed APIs? vscode's textmate grammars are capable of specifying tokens which span multiple lines, so I feel like I'm missing something :/

**Comment by hediet** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!

**Comment by dhakehurst** [comment]
Hi,
no I do not agree :)

as I stated above, I want to tokenise a whole sentence, irrespective of where line breaks are.

Is it such a major change in architecture to enable non-line-based token provider, I would not have thought so.
